README.md
# NodeJS [FKWallet](https://www.fkwallet.ru/docs) API

[![license](https://img.shields.io/npm/l/fkwallet?style=flat-square)](https://github.com/xTCry/node-fkwallet-api/blob/main/LICENSE)
[![npm](https://img.shields.io/npm/v/fkwallet?style=flat-square)](https://npmjs.com/package/fkwallet)
[![downloads](https://img.shields.io/npm/dm/fkwallet?style=flat-square)](https://npmjs.com/package/fkwallet)
[![GitHub](https://img.shields.io/github/stars/xTCry/node-fkwallet-api?style=flat-square)](https://github.com/xTCry/node-fkwallet-api)
[![last commit](https://img.shields.io/github/last-commit/xTCry/node-fkwallet-api?style=flat-square)](https://github.com/xTCry/node-fkwallet-api)

## 📦 Установка

```shell
yarn add fkwallet
```

## 🛠️ Использование

```typescript
import 'dotenv/config';
import { ApiClient, CryptoType } from 'fkwallet';

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

const main = async () => {
  const api = new ApiClient(
    process.env.WALLET_PROVIDER_FKWALLET_SHOP_ID!,
    process.env.WALLET_PROVIDER_FKWALLET_API_KEY!,
  );

  let time = Date.now();
  const balance = await api.getBalance();
  console.log('balance', balance);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(1e3);
  time = Date.now();
  const btcAddressNew = await api.createCryptoAddress(CryptoType.BTC);
  console.log('btcAddressNew', btcAddressNew);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(1e3);
  time = Date.now();
  const btcAddress = await api.getCryptoAddress(CryptoType.BTC);
  console.log('btcAddress', btcAddress);
  console.log(`Time: ${Date.now() - time} ms`);

  await sleep(1e3);
  time = Date.now();
  const paymentStatus = await api.getPaymentStatus({ payment_id: 22664540 });
  console.log('paymentStatus', paymentStatus.data.payment_id);
  console.log(`Time: ${Date.now() - time} ms`);
};
main().then();

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/xTCry/node-fkwallet-api/compare/v1.0.1...v1.0.2) (2022-09-19)


### Bug Fixes

* **api:** fixed typo in sign for `Transfer` ([c2275c3](https://github.com/xTCry/node-fkwallet-api/commit/c2275c3149b9268699b7e9eb7c8f45ba59fcef77))


### Others

* **types:** updated optional parameters ([5007d6b](https://github.com/xTCry/node-fkwallet-api/commit/5007d6bd835431ec2cc71fe76fe03fcc29a32473))

### [1.0.1](https://github.com/xTCry/node-fkwallet-api/compare/v1.0.0...v1.0.1) (2021-11-07)


### Bug Fixes

* **sign:** fixed check signs body field names ([88a3c22](https://github.com/xTCry/node-fkwallet-api/commit/88a3c228edcc23256b1cef87d7b94605f752470c))


### Others

* **.versionrc:** updated standard-version types ([59dc520](https://github.com/xTCry/node-fkwallet-api/commit/59dc52080f64ea08040d50d21146bf7cbbd7ab01))

## 1.0.0 (2021-11-07)


src/types.ts
export enum ErrorType {
  InvalidPurse = 'InvalidPurse',
  InvalidCard = 'InvalidCard',
  BalanceToLow = 'BalanceToLow',
  CoolDown = 'CoolDown',
  Wait24Hour = 'Wait24Hour',
  Unknown = 'Unknown',
}

export enum ActionType {
  GetBalance = 'get_balance',
  Cashout = 'cashout',
  GetPaymentStatus = 'get_payment_status',
  Transfer = 'transfer',
  OnlinePayment = 'online_payment',
  Providers = 'providers',
  CheckOnlinePayment = 'check_online_payment',

  CreateCryptoAddress_BTC = 'create_btc_address',
  CreateCryptoAddress_LTC = 'create_ltc_address',
  CreateCryptoAddress_ETH = 'create_eth_address',
  GetCryptoAddress_BTC = 'get_btc_address',
  GetCryptoAddress_LTC = 'get_ltc_address',
  GetCryptoAddress_ETH = 'get_eth_address',
  GetCryptoTransaction_BTC = 'get_btc_transaction',
  GetCryptoTransaction_LTC = 'get_ltc_transaction',
  GetCryptoTransaction_ETH = 'get_eth_transaction',
}

export enum CryptoActionType {
  CreateCryptoAddress = 'create_crypto_address',
  GetCryptoAddress = 'get_crypto_address',
  GetCryptoTransaction = 'get_crypto_transaction',
}

export enum CryptoType {
  BTC = 'btc',
  LTC = 'ltc',
  ETH = 'eth',
}

export const cryptoActions: Record<CryptoType, Record<CryptoActionType, ActionType>> = {
  [CryptoType.BTC]: {
    [CryptoActionType.CreateCryptoAddress]: ActionType.CreateCryptoAddress_BTC,
    [CryptoActionType.GetCryptoAddress]: ActionType.GetCryptoAddress_BTC,
    [CryptoActionType.GetCryptoTransaction]: ActionType.GetCryptoTransaction_BTC,
  },
  [CryptoType.LTC]: {
    [CryptoActionType.CreateCryptoAddress]: ActionType.CreateCryptoAddress_LTC,
    [CryptoActionType.GetCryptoAddress]: ActionType.GetCryptoAddress_LTC,
    [CryptoActionType.GetCryptoTransaction]: ActionType.GetCryptoTransaction_LTC,
  },
  [CryptoType.ETH]: {
    [CryptoActionType.CreateCryptoAddress]: ActionType.CreateCryptoAddress_ETH,
    [CryptoActionType.GetCryptoAddress]: ActionType.GetCryptoAddress_ETH,
    [CryptoActionType.GetCryptoTransaction]: ActionType.GetCryptoTransaction_ETH,
  },
};

/** Список доступных валют */
export enum CurrencyType {
  /**
   * WebMoney WMZ *(комиссия 7.00%)*
   *
   * `Z12345678912` **Мин - 10 руб, Макс - 15000 руб**
   */
  WebMoney_WMZ = 2,
  /**
   * QIWI кошелек *(комиссия 4.00%)*
   *
   * `+79261234567` **Мин - 100 руб, Макс - 15000 руб**
   */
  QIWI = 63,
  /**
   * Perfect Money USD *(комиссия 8.00%)*
   *
   * `U1234567` **Мин - 10 руб, Макс - 15000 руб**
   */
  Perfect_Money_USD = 64,
  /**
   * Яндекс.Деньги *(комиссия 3.00%)*
   *
   * `410011812345678` **Мин - 10 руб, Макс - 15000 руб**
   */
  YooMoney = 45,
  /**
   * VISA/MASTERCARD UAH *(комиссия 2.50%)*
   *
   * `4012888812345678` **Мин - 272 руб, Макс - 27227 руб**
   * `4012888812345678` **Мин - 100 UAH, Макс - 10000 UAH**
   */
  Card_EAH = 67,
  /**
   * Perfect Money EUR *(комиссия 6.50%)*
   *
   * `U1234567` **Мин - 10 руб, Макс - 15000 руб**
   */
  Perfect_Money_EUR = 69,
  /**
   * PayPal *(комиссия 3.50%)*
   *
   * `user@site.ru` **Мин - 10 руб, Макс - 15000 руб**
   */
  PayPal = 70,
  /**
   * Мобильный Платеж Мегафон *(комиссия 1.00%)*
   *
   * `+79261234567` **Мин - 10 руб, Макс - 15000 руб**
   */
  Mobile_Megafon = 82,
  /**
   * Мобильный Платеж Билайн *(комиссия 1.00%)*
   *
   * `+79261234567` **Мин - 10 руб, Макс - 15000 руб**
   */
  Mobile_Beeline = 83,
  /**
   * обильный Платеж МТС *(комиссия 1.00%)*
   *
   * `+79261234567` **Мин - 10 руб, Макс - 15000 руб**
   */
  Mobile_MTS = 84,
  /**
   * VISA/MASTERCARD RUB *(комиссия 4.00%)*
   * `4012888812345678` **Мин - 100 руб, Макс - 10000 руб**
   */
  Card_RUB = 94,
  /**
   * Мобильный Платеж Tele2 *(комиссия 1.00%)*
   *
   * `+79261234567` **Мин - 10 руб, Макс - 15000 руб**
   */
  Mobile_Tele2 = 132,
  /**
   * PAYEER USD *(комиссия 4.00%)*
   *
   * `P12345678` **Мин - 712 руб, Макс - 35610 руб**
   * `P12345678` **Мин - 10 USD, Макс - 500 USD**
   */
  PAYEER_USD = 115,
  /**
   * PAYEER RUB *(комиссия 5.50%)*
   *
   * `P12345678` **Мин - 10 руб, Макс - 100000 руб**
   */
  PAYEER_RUB = 114,
  /**
   * FK WALLET RUB *(комиссия 0.00%)*
   *
   * `F123456789` **Мин - 10 руб, Макс - 600000 руб**
   */
  FK_WALLET_RUB = 133,
  /**
   * ADVCASH USD *(комиссия 8.00%)*
   *
   * `123456789` **Мин - 10 руб, Макс - 15000 руб**
   */
  ADVCASH_USD = 136,
  /**
   * ADVCASH RUB *(комиссия 5.00%)*
   *
   * `123456789` **Мин - 10 руб, Макс - 15000 руб**
   */
  ADVCASH_RUB = 150,
  /**
   * QIWI EURO *(комиссия 4.00%)*
   *
   * `+79261234567` **Мин - 100 руб, Макс - 15000 руб**
   */
  QIWI_EURO = 161,
  /**
   * VISA/MASTERCARD KZT *(комиссия 3.00%)*
   *
   * `4012888812345678` **Мин - 166 руб, Макс - 83426 руб**
   * `4012888812345678` **Мин - 1000 KZT, Макс - 500000 KZT**
   */
  Card_KTZ = 186,
}

/** Комиссия от типа валюты *(через какое-то время может измениться)* */
export const commissions: Record<CurrencyType, number> = {
  [CurrencyType.WebMoney_WMZ]: 0.07,
  [CurrencyType.QIWI]: 0.04,
  [CurrencyType.Perfect_Money_USD]: 0.08,
  [CurrencyType.YooMoney]: 0.03,
  [CurrencyType.Card_EAH]: 0.025,
  [CurrencyType.Perfect_Money_EUR]: 0.065,
  [CurrencyType.PayPal]: 0.035,
  [CurrencyType.Mobile_Megafon]: 0.01,
  [CurrencyType.Mobile_Beeline]: 0.01,
  [CurrencyType.Mobile_MTS]: 0.01,
  [CurrencyType.Card_RUB]: 0.04,
  [CurrencyType.Mobile_Tele2]: 0.01,
  [CurrencyType.PAYEER_USD]: 0.04,
  [CurrencyType.PAYEER_RUB]: 0.055,
  [CurrencyType.FK_WALLET_RUB]: 0.0,
  [CurrencyType.ADVCASH_USD]: 0.08,
  [CurrencyType.ADVCASH_RUB]: 0.05,
  [CurrencyType.QIWI_EURO]: 0.04,
  [CurrencyType.Card_KTZ]: 0.03,
};

/** Статус операции вывода из кошелька */
export enum PaymentType {
  /** Новая */
  New = 'New',
  /** В процессе */
  InProcess = 'In process',
  /** Выполнена */
  Completed = 'Completed',
  /** Отменена */
  Canceled = 'Canceled',
}

// Withdrawal

/** Параметры уведомления о выводе */
export type PaymentParams = {
  /** Ваш номер кошелька */
  wallet_id: string;
  /** Номер операции */
  order_id: number;
  /** Статус операции */
  status: PaymentStatusType;
  /** Сумма */
  amount: number;
  /** Ваш номер операции */
  user_order_id: number;
  /**
   * Контрольная подпись MD5, формируется из параметров
   * `wallet_id`, `order_id`, `user_order_id`, `status`, `amount` и `API ключа`, например
   *
   * ```js
   * md5('F123456789'.'123'.'321'.'1'.'100.12'.'JG68NC68DJC8S0DCH6')
   * ```
   */
  sign: string;
};

/** Статус выплаты */
export enum PaymentStatusType {
  /** Успешно выполнен */
  Success = 1,
  /** В процессе */
  Process = 7,
  /** Ошибка */
  Error = 9,
}

// Crypto transaction

/** Уведомление о новой крипто транзакции */
export type CryptoTransactionParams = {
  /**	Ваш номер кошелька */
  wallet_id: string;
  /**	Адрес */
  address: string;
  /**	ID транзакции */
  transaction_id: string;
  /**	Сумма */
  amount: string;
  /**	Комиссия */
  fee: string;
  /**	Количество подтверждений */
  confirmations: string;
  /**	Дата */
  date: string;
  /**
   * Контрольная подпись MD5, формируется из параметров
   * `wallet_id`, `address`, `transaction_id`, `amount`, `fee`, `confirmations`, `date` и `API ключа`, например
   *
   * ```js
   * md5('F123456789'.'567897688tgydscsd8985'.'gb56yu3txdy237dy2xu8d2983tdxy23dux2873d7yx20d'.'0.001'.'0'.'0'.'2017-01-01 23:32:33'.'JG68NC68DJC8S0DCH6')
   * ```
   */
  sign: string;
};

// Responses

export interface IResponseData<T> {
  status: 'info' | 'error' | /* ? */ 'success';
  desc: string;
  data: T;
}

export type ResponseBalance = Record<string, string>;

// Cashout

export type ParamsCashout = {
  /** Кошелек для вывода */
  purse: string;
  /** Сумма вывода */
  amount: number;
  /** Примечание */
  desc?: string;
  /** Валюта для вывода */
  currency: CurrencyType;
  /** Отключить автоматический обмен */
  disable_exchange?: number | boolean;
  /** Ваш номер операции (целое положительное число) */
  order_id?: number;
  /**
   * При указании номера операции с данным параметром, будет выдана ошибка,
   * если заявка с таким номером операции уже существует
   */
  check_duplicate?: number | boolean;
};

export type ResponseCashout = { payment_id: string };

// GetPaymentStatus

export type ParamsGetPaymentStatus = {
  /** Номер операции полученный при отправке запроса */
  payment_id?: number;
  /** **или** ваш номер операции */
  user_order_id?: number;
};

export type ResponseGetPaymentStatus = {
  payment_id?: string;
  user_order_id?: number;
  status: PaymentType;
};

// Transfer

export type ParamsTransfer = {
  /** Кошелек для перевода */
  purse: string;
  /** Сумма перевода */
  amount: number;
};

export type ResponseTransfer = undefined;

// OnlinePayment

export type ParamsOnlinePayment = {
  /** ID услуги */
  service_id: number;
  /** Кошелек для перевода */
  account: string;
  /** Сумма перевода */
  amount: number;
  /** Ваш номер операции (целое положительное число) */
  order_id?: number;
  /**
   * При указании номера операции с данным параметром, будет выдана ошибка,
   * если заявка с таким номером операции уже существует
   */
  check_duplicate?: number | boolean;
};

export type ResponseOnlinePayment = {
  payment_id: string;
};

// Providers

export type ResponseProviders = {
  /** Ид сервиса */
  id: string;
  /**
   * Название сервиса.
   *
   * Например, `"МТС"`, `"Билайн"`
   */
  name: string;
  /**
   * Минимальная сумма перевода
   *
   * Например, `"5.00"`
   */
  min_amount: string;
  /**
   * Комиссия перевода
   *
   * Например, `"0"`
   */
  commission: string;
}[];

// CheckOnlinePayment

export type ParamsCheckOnlinePayment = {
  /** Номер платежа, полученный при отправке */
  payment_id?: number;
  /** **или** ваш номер операции */
  user_order_id?: number;
};

export type ResponseCheckOnlinePayment = {
  payment_id?: string;
  user_order_id?: number;
  status: PaymentType;
};

// CreateCryptoAddress

export type ResponseCreateCryptoAddress = {
  address: string;
};

// GetCryptoAddress

export type ResponseGetCryptoAddress = {
  address: string;
  valid: string;
};

// GetCryptoAddress

export type ResponseGetCryptoTransaction = {
  address: string;
  transaction_id: string;
  amount: string;
  fee: string;
  confirmations: string;
  date: string;
}[];


src/errors.ts
import { ErrorType, IResponseData } from './types';

const getErrorType = ({ desc }: IResponseData<undefined>) => {
  const lowerDesc = desc.toLowerCase();
  switch (true) {
    case lowerDesc.startsWith('Invalid Purse '.toLowerCase()):
      return ErrorType.InvalidPurse;
    case /Only ([a-z]{,10}) cards/i.test(lowerDesc):
      return ErrorType.InvalidCard;
    // TODO: check it
    case lowerDesc.startsWith('Balance to low'.toLowerCase()):
      return ErrorType.BalanceToLow;
    case lowerDesc.startsWith('Слишком частые запросы к API'.toLowerCase()):
      return ErrorType.CoolDown;
    case lowerDesc.startsWith('Адрес можно продлить не ранее, чем через 24 часа с момента резерва'.toLowerCase()):
      return ErrorType.Wait24Hour;

    // TODO: add other errors

    default:
      return ErrorType.Unknown;
  }
};

export class ResponseError extends Error {
  name = this.constructor.name;

  public readonly type: ErrorType;

  constructor(readonly body: IResponseData<any>) {
    super();
    this.message = body.desc;
    this.type = getErrorType(body);
  }

  get [Symbol.toStringTag](): string {
    return `[Error ${this.name}; type=${this.type}] ${this.message}`;
  }

  toJSON(): Error & { type: ErrorType } {
    return {
      message: this.message,
      name: this.name,
      type: this.type,
      stack: this.stack,
    };
  }
}


src/api-client.ts
import axios, { AxiosInstance, AxiosRequestConfig, Method } from 'axios';
import { stringify } from 'query-string';
import * as types from './types';
import { ResponseError } from './errors';
import { md5 } from './utils';

// TODO: сделать нормализацию возвращаемых параметров к правильным типам

export class ApiClient {
  public static readonly API_URL = 'https://www.fkwallet.ru/api_v1.php';

  private readonly _httpClient: AxiosInstance;

  constructor(public readonly walletId: string, public readonly apiKey: string, apiOptions: AxiosRequestConfig = {}) {
    this._httpClient = axios.create({
      baseURL: ApiClient.API_URL,
      timeout: 10e3,
      ...apiOptions,
    });

    // this._httpClient.interceptors.request.use((config) => {
    //   console.log('config', config);
    //   return config;
    // });
  }

  public async _callApi<T = unknown>(action: types.ActionType, data: any = {}, method: Method = 'POST') {
    try {
      const resp = await this._httpClient.request<types.IResponseData<T>>({
        method,
        data: stringify({
          wallet_id: this.walletId,
          ...data,
          action,
          sign: this.getSign(action, data),
        }),
      });
      if (resp.data.status === 'error') {
        throw new ResponseError(resp.data);
      }
      return resp.data;
    } catch (e) {
      if ((e as any as types.IResponseData<any>).status === 'error') {
        throw new ResponseError(e as types.IResponseData<any>);
      }
      throw e;
    }
  }

  public getSign(action: types.ActionType, payload: any) {
    switch (action) {
      case types.ActionType.GetBalance:
      case types.ActionType.Providers:
      case types.ActionType.CreateCryptoAddress_BTC:
      case types.ActionType.CreateCryptoAddress_LTC:
      case types.ActionType.CreateCryptoAddress_ETH:
      case types.ActionType.GetCryptoAddress_BTC:
      case types.ActionType.GetCryptoAddress_LTC:
      case types.ActionType.GetCryptoAddress_ETH:
        return md5(`${this.walletId}${this.apiKey}`);
      case types.ActionType.Cashout:
        return md5(`${this.walletId}${payload.currency}${payload.amount}${payload.purse}${this.apiKey}`);
      case types.ActionType.GetPaymentStatus:
      case types.ActionType.CheckOnlinePayment:
        return md5(`${this.walletId}${payload.payment_id || payload.user_order_id}${this.apiKey}`);
      case types.ActionType.Transfer:
        return md5(`${this.walletId}${payload.amount}${payload.purse}${this.apiKey}`);
      case types.ActionType.OnlinePayment:
        return md5(`${this.walletId}${payload.amount}${payload.account}${this.apiKey}`);
      default:
        return null;
    }
  }

  /**
   * Проверяет подпись уведомления о выплате
   * @param body Объект уведомления
   * @returns {boolean} Признак валидности
   */
  public checkSignPayment(body: types.PaymentParams) {
    const compareSign = md5(
      `${this.walletId}${body.order_id}${body.user_order_id}${body.status}${body.amount}${this.apiKey}`,
    );

    return body.sign === compareSign;
  }

  /**
   * Проверяет подпись уведомления о новой крипто транзакции
   * @param body Объект уведомления
   * @returns {boolean} Признак валидности
   */
  public checkSignCryptoTransaction(body: types.CryptoTransactionParams) {
    const compareSign = md5(
      `${this.walletId}${body.address}${body.transaction_id}${body.amount}${body.fee}${body.confirmations}${body.date}${this.apiKey}`,
    );

    return body.sign === compareSign;
  }

  /**
   * Получает баланс кошелька
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Wallet balance","data":{"RUB":"50.00","USD":"0.00","EUR":"0.00"}}
   * ```
   */
  public async getBalance() {
    return await this._callApi<types.ResponseBalance>(types.ActionType.GetBalance);
  }

  /**
   * Выводит средства из кошелька
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Payment send","data":{"payment_id":"543273"}}
   * ```
   */
  public async cashout(params: types.ParamsCashout) {
    return await this._callApi<types.ResponseCashout>(types.ActionType.Cashout, params);
  }

  /**
   * Получает статус операции вывода из кошелька
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Order status","data":{"payment_id":"543273","status":"Canceled"}}
   * ```
   */
  public async getPaymentStatus(params: types.ParamsGetPaymentStatus) {
    return await this._callApi<types.ResponseGetPaymentStatus>(types.ActionType.GetPaymentStatus, params);
  }

  /**
   * Переводит на другой кошелек
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Payment send"}
   * ```
   */
  public async transfer(params: types.ParamsTransfer) {
    return await this._callApi<types.ResponseTransfer>(types.ActionType.Transfer, params);
  }

  /**
   * Оплата онлайн услуг
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Payment send","data":{"payment_id":"543273"}}
   * ```
   */
  public async onlinePayment(params: types.ParamsOnlinePayment) {
    return await this._callApi<types.ResponseOnlinePayment>(types.ActionType.OnlinePayment, params);
  }

  /**
   * Получает список сервисов для онлайн оплаты
   *
   * *Пример возвращаемых данных*
   * ```json
   * {
   *   "status": "info",
   *   "desc": "Providers list",
   *   "data": [
   *     {
   *       "id": "1",
   *       "name": "МТС",
   *       "min_amount": "5.00",
   *       "commission": "0"
   *     },
   *     {
   *       "id": "2",
   *       "name": "Билайн",
   *       "min_amount": "5.00",
   *       "commission": "0"
   *     }
   *   ]
   * }
   * ```
   */
  public async providers() {
    return await this._callApi<types.ResponseProviders>(types.ActionType.Providers);
  }

  /**
   * Проверяет статус онлайн платежа
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Order status","data":{"payment_id":"6532323","status":"Canceled"}}
   * ```
   */
  public async checkOnlinePayment(params: types.ParamsCheckOnlinePayment) {
    return await this._callApi<types.ResponseCheckOnlinePayment>(types.ActionType.CheckOnlinePayment, params);
  }

  /**
   * Создает крипто адрес
   *
   * **Адрес резервируется на 3 дня, для продления резерва необходимо повторно отправить данный запрос**
   * *(не ранее, чем через 24 часа с момента резерва).*
   *
   * *Пожалуйста не спамьте запросами, достаточно отправлять запрос раз в 2 дня, чтобы адрес постоянно был зарезервирован за вами*
   *
   * *Пример возвращаемых данных*
   * ```json
   * {"status":"info","desc":"Address created","data": {"address": "4eftk98j9h76g5454er5ty8uh3dwec"}}
   * ```
   */
  public async createCryptoAddress(crypo: types.CryptoType) {
    return await this._callApi<types.ResponseCreateCryptoAddress>(
      types.cryptoActions[crypo][types.CryptoActionType.CreateCryptoAddress],
    );
  }

  /**
   * Получает крипто адрес
   *
   * *Пример возвращаемых данных*
   * ```json
   * {
   *   "status": "info",
   *   "desc": "Адрес 46D98a09Fd204C9C72d2A2Dd3563fF5495aD41D0 успешно зарезервирован до 2009-01-03 21:15:42",
   *   "data": {
   *     "address": "46D98a09Fd204C9C72d2A2Dd3563fF5495aD41D0",
   *     "valid": "2020-05-16 14:54:52"
   *   }
   * }
   * ```
   */
  public async getCryptoAddress(crypo: types.CryptoType) {
    return await this._callApi<types.ResponseGetCryptoAddress>(
      types.cryptoActions[crypo][types.CryptoActionType.GetCryptoAddress],
    );
  }

  /**
   * Получает информацию по крипто транзакции
   *
   * *Пример возвращаемых данных*
   * ```json
   * {
   *   "status": "info",
   *   "desc": "",
   *   "data": [
   *     {
   *       "address": "4eftk98j9h76g5454er5ty8uh3dwec",
   *       "transaction_id": "gb56yu3txdy237dy2xu8d2983tdxy23dux2873d7yx20d",
   *       "amount": "0.001",
   *       "fee": "0",
   *       "confirmations": "4",
   *       "date": "2017-01-01 23:32:33"
   *     }
   *   ]
   * }
   * ```
   */
  public async getCryptoTransaction(crypo: types.CryptoType) {
    return await this._callApi<types.ResponseGetCryptoTransaction>(
      types.cryptoActions[crypo][types.CryptoActionType.GetCryptoTransaction],
    );
  }
}
